python/ray/dag/compiled_dag_node.py

```diff
@@ -868,6 +876,158 @@ def _get_node_id(self):
+            dag.with_type_hint(TorchTensorType(transport="nccl"))
+            dag = a.no_op.bind(dag)
+        dag.experimental_compile()
+
+        In the above example, communication between a.no_op occurs via an NCCL
+        channel, while communication between the driver process and a.no_op occurs
+        via shared memory channels. The example experiences a deadlock because the
+        completion of the write function in the first a.no_op requires the second
+        a.no_op to simultaneously call the read function. However, each actor has
+        a list of tasks, each assigned a bind index, and these tasks are executed
+        sequentially in ascending order of their bind index on the actor. Therefore,
+        it’s impossible for both writer and reader on the same actor to write and
+        read simultaneously.
+
+        We can create a depends-on graph based on the above rules. Then, the graph
+        will look like this:
+
+                              |---|
+                              |   v
+        driver -> a.no_op -> a.no_op -> driver
+
+        Then, we use topological sort to verify whether the graph has a cycle.
+
+        If you are interested in the detailed explanation, please refer to
+        https://github.com/ray-project/ray/pull/45960.
+        """
+        assert self.idx_to_task
+        assert self.actor_to_tasks
+
+        class GraphNode:
+            def __init__(self):
+                self.in_edges = set()
+                self.out_edges = set()
+
+            @property
+            def in_degree(self) -> int:
+                return len(self.in_edges)
+
+        from ray.dag import ClassMethodNode
+
+        def _get_next_task_idx(task: "CompiledTask") -> Optional[int]:
+            if not isinstance(task.dag_node, ClassMethodNode):
+                return None
+            actor_handle = task.dag_node._get_actor_handle()
+            bind_index = task.dag_node._get_bind_index()
+            for same_node_task in self.actor_to_tasks[actor_handle]:
+                if same_node_task.dag_node._get_bind_index() == bind_index + 1:
+                    return same_node_task.idx
+            return None
+
+        def _add_edge(
+            graph: Dict[int, GraphNode], from_idx: int, to_idx: Optional[int]
+        ):
+            if to_idx is None:
+                return
+            graph[from_idx].out_edges.add(to_idx)
+            graph[to_idx].in_edges.add(from_idx)
+
+        graph = defaultdict(GraphNode)
+        for idx, task in self.idx_to_task.items():
+            # Add an edge from task_{bind_index} to task_{bind_index+1}
+            # on the same actor.
+            next_task_idx = _get_next_task_idx(task)
+            _add_edge(graph, idx, next_task_idx)
+            for downstream_idx in task.downstream_node_idxs:
+                if task.dag_node.type_hint.requires_nccl():
+                    # Add an edge from the writer and reader of an NCCL channel
```

Comment:
This seems to conflict with the language in the comment above and in the Google doc. The comment and Google doc say that an edge is only added from the reader, but this comment here says that an edge is added from both the writer and the reader.

Reply:
Good catch. I forgot to update it. They are equivalent, and we will still add an edge from the writer to its next_bind_index by following rule 1. I will update it.

Reply:
Updated https://github.com/ray-project/ray/pull/45960/commits/f7c73d5d49bb31570bfbdf1702d5ea91aacb6a25